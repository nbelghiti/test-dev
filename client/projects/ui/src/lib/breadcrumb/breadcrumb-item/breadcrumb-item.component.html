<div uiFlyout *ngIf="items && items.length > 1; else asLink">
	<button
		uiFlyoutToggle
		class="c-breadcrumb-item"
		[ngClass]="{
			'c-breadcrumb-item--first': first
		}"
		[title]="value"
	>
		<span class="c-breadcrumb-item__label">
			{{label}}
		</span>

		<div class="c-breadcrumb-item__item">
			<span
				*ngIf="conflict"
				class="c-breadcrumb-item__icon"
			>
				<ui-icon
					[name]="IconName.exclamationCircle"
					[size]="IconSize.tiny"
					[type]="IconType.error"
				>
				</ui-icon>
			</span>

			<span class="c-breadcrumb-item__value">
				{{value}}
			</span>

			<span class="c-breadcrumb-item__arrow">
				<ui-icon
					[name]="IconName.caretDown"
					[size]="IconSize.small"
				>
				</ui-icon>
			</span>
		</div>
	</button>

	<ui-filter-menu
		uiFlyoutContent
		class="c-flyout__content--left"
		label="value"
		[items]="items"
		[formControl]="activeBreadcrumb"
	>
	</ui-filter-menu>
</div>

<ng-template #asLink>
	<a
		class="c-breadcrumb-item"
		[ngClass]="{
			'c-breadcrumb-item--first': first
		}"
		[routerLink]="link"
		[queryParams]="queryParams"
		[title]="value"
	>
		<span class="c-breadcrumb-item__label">
			{{label}}
		</span>

		<div class="c-breadcrumb-item__item">
			<span
				*ngIf="conflict"
				class="c-breadcrumb-item__icon"
			>
				<ui-icon
					[name]="IconName.exclamationCircle"
					[size]="IconSize.tiny"
					[type]="IconType.error"
				>
				</ui-icon>
			</span>

			<span class="c-breadcrumb-item__value">
				{{value}}
			</span>
		</div>
	</a>
</ng-template>
