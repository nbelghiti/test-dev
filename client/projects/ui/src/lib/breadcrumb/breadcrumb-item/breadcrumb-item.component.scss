// =============================================================================
// :: Imports
// =============================================================================
@import "global";

// =============================================================================
// :: Breadcrumb item
// =============================================================================
.c-breadcrumb-item {
	position: relative;

	display: inline-flex;
	flex-direction: column;
	padding: 1.4rem 2rem;

	text-decoration: none;

	border-right: $border-width-base solid $border-color-grey-86;
	color: $text-color-grey-46;

	&:hover {
		color: $text-color-base;

		cursor: pointer;
	}
}

// =============================================================================
// :: Elements
// =============================================================================
.c-breadcrumb-item__item {
	display: inline-flex;
	align-items: center;

	margin-top: 0.2rem;
}

.c-breadcrumb-item__label {
	@include extend-text-overflow();

	display: inline-block;
	max-width: 20rem;

	font-size: $font-size-2x-small;
	line-height: $line-height-normalize;
	text-align: left;
	text-transform: uppercase;

	white-space: nowrap;
}

.c-breadcrumb-item__value {
	@include extend-text-overflow();

	display: inline-block;
	max-width: 20rem;

	font-size: $font-size-1x-small;
	font-weight: $font-weight-bold;
	line-height: $line-height-base;

	white-space: nowrap;
}

.c-breadcrumb-item__arrow {
	display: inline-block;
	margin-left: 1rem;
}

.c-breadcrumb-item__icon {
	display: inline-block;
	margin-right: 0.4rem;
}

// =============================================================================
// :: Modifiers
// =============================================================================
.c-breadcrumb-item--first {
	background-color: $background-color-grey-96;
}
