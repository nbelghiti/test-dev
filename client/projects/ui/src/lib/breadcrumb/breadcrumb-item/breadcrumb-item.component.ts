import { Component, Input, OnChanges, SimpleChanges } from '@angular/core';
import { FormControl } from '@angular/forms';

import { IFilterMenuItem } from '../../filter/filter-menu/filter-menu.component.types';
import { IconName, IconSize, IconType } from '../../icon/icon.types';

@Component({
	selector: 'ui-breadcrumb-item',
	templateUrl: './breadcrumb-item.component.html',
	styleUrls: ['./breadcrumb-item.component.scss']
})
export class BreadcrumbItemComponent implements OnChanges {
	@Input() public label: string = '';
	@Input() public value: string = '';
	@Input() public items: IFilterMenuItem[] = null;
	@Input() public first?: boolean = false;
	@Input() public link?: string[];
	@Input() public conflict?: boolean = false;
	@Input() public queryParams?: { [key: string]: string };

	public IconName: typeof IconName = IconName;
	public IconSize: typeof IconSize = IconSize;
	public IconType: typeof IconType = IconType;
	public activeBreadcrumb: FormControl = new FormControl();

	public ngOnChanges(changes: SimpleChanges): void {
		if (changes.value && changes.value.currentValue) {
			this.activeBreadcrumb.setValue(
				{ value: changes.value.currentValue },
				{ emitEvent: false }
			);
		}
	}
}
