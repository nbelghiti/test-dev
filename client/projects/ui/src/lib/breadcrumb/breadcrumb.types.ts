import { IFilterMenuItem } from '../filter/filter-menu/filter-menu.component.types';

export interface IBreadcrumbItem {
	value: string;
	label?: string;
	link?: string[];
	queryParams?: { [key: string]: string };
	items?: IFilterMenuItem[];
}
