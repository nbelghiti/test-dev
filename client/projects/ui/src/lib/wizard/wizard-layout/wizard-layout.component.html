<header
	class="l-step-wizard__header"
	[class.l-step-wizard__header--with-breadcrumb]="breadcrumb.length > 0"
>
	<div class="l-step-wizard__header-content l-container l-container--fluid">
		<ng-content select="header"></ng-content>
	</div>

	<div
		*ngIf="breadcrumb?.length > 0"
		class="l-step-wizard__header-breadcrumb"
	>
		<ui-breadcrumb [items]="breadcrumb"></ui-breadcrumb>
	</div>
</header>


<div
	class="l-step-wizard__scroll-container"
	[class.u-padding-none]="!withPadding"
>
	<div
		class="l-step-wizard__container"
		[class.l-container]="withPadding"
		[class.l-container--fluid]="withPadding"
	>
		<div
			class="l-step-wizard__content"
			[ngClass]="{
				'l-step-wizard__content--standalone': !standalone,
				'l-step-wizard__content--full-width': isFullScreen
			}"
		>
			<ng-content select="body"></ng-content>
		</div>

		<ui-steps
			class="l-step-wizard__steps"
			*ngIf="showSteps"
			[activeIndex]="activeIndex"
			[steps]="steps"
		></ui-steps>
	</div>
</div>

<footer class="l-step-wizard__footer">
	<div
		class="l-container l-container--fluid"
		[class.l-step-wizard__footer-content]="!hideFooter"
	>
		<ng-content select="footer"></ng-content>
	</div>
</footer>
