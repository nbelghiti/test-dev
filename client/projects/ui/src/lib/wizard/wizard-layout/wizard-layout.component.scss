// =============================================================================
// :: Imports
// =============================================================================
@import "global";

// =============================================================================
// :: Wizard layout
// =============================================================================

.l-step-wizard__header-breadcrumb {
	border-top: $border-width-base solid $border-color-grey-86;
}
