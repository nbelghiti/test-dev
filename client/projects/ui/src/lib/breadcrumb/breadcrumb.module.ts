import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { TranslateModule } from '@ngx-translate/core';

import { ButtonModule } from '../button/button.module';
import { FilterModule } from '../filter/filter.module';
import { FlyoutModule } from '../flyout/flyout.module';
import { IconModule } from '../icon/icon.module';

import { BreadcrumbItemComponent } from './breadcrumb-item/breadcrumb-item.component';
import { BreadcrumbComponent } from './breadcrumb/breadcrumb.component';

@NgModule({
	imports: [
		CommonModule,
		IconModule,
		ButtonModule,
		RouterModule,
		FlyoutModule,
		FilterModule,
		ReactiveFormsModule,
		TranslateModule
	],
	declarations: [
		BreadcrumbItemComponent,
		BreadcrumbComponent,
	],
	exports: [
		BreadcrumbItemComponent,
		BreadcrumbComponent,
	],
})
export class BreadcrumbModule {}
