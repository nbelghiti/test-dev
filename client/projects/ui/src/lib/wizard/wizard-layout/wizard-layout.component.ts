import { Component, HostBinding, Input } from '@angular/core';

import { IBreadcrumbItem } from '../../breadcrumb/breadcrumb.types';
import { IStep } from '../../steps/steps.types';

@Component({
	selector: 'ui-wizard-layout',
	templateUrl: './wizard-layout.component.html',
	styleUrls: [
		'./wizard-layout.component.scss',
	],
})
export class WizardLayoutComponent {
	@HostBinding('class') class = 'l-step-wizard';
	@HostBinding('class.l-wizard--cover')
	get isCover(): boolean {
		return this.isFullScreen;
	}

	@Input() public activeIndex: number;
	@Input() public breadcrumb: IBreadcrumbItem[] = [];
	@Input() public hideFooter: boolean = false;
	@Input() public showSteps: boolean = false;
	@Input() public standalone: boolean = false;
	@Input() public withPadding: boolean = true;
	@Input() public isFullScreen: boolean = false;
	@Input() public steps: IStep[];
}
