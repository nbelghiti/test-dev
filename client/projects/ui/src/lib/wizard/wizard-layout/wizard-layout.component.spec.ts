import { Component, DebugElement } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { By } from '@angular/platform-browser';
import { RouterTestingModule } from '@angular/router/testing';
import { TranslateModule } from '@ngx-translate/core';

import { BreadcrumbItemComponent } from '../../breadcrumb/breadcrumb-item/breadcrumb-item.component';
import { BreadcrumbComponent } from '../../breadcrumb/breadcrumb/breadcrumb.component';
import { ButtonComponent } from '../../button/button/button.component';
import { FilterMenuItemComponent } from '../../filter/filter-menu-item/filter-menu-item.component';
import { FilterMenuComponent } from '../../filter/filter-menu/filter-menu.component';
import { IconComponent } from '../../icon/icon.component';
import { StepsModule } from '../../steps/steps.module';
import { IStep } from '../../steps/steps.types';

import { WizardLayoutComponent } from './wizard-layout.component';

const MockSteps = [{
	title: 'First step',
	description: 'This is the first step',
}, {
	title: 'Second step',
	description: 'This is the second step',
}, {
	title: 'Third step',
	description: 'This is the third step',
}];

@Component({
	template: `
		<ui-wizard-layout [showSteps]="showSteps" [steps]="steps" [activeIndex]="activeIndex">
			<ng-container ngProjectAs="header">
				<h1 id="test-header">Test header</h1>
			</ng-container>
			<ng-container ngProjectAs="body">
				<p id="test-body">Test body</p>
			</ng-container>
			<ng-container ngProjectAs="footer">
				<div id="test-footer">Test footer</div>
			</ng-container>
		</ui-wizard-layout>
	`
})
export class TestWrapperComponent {
	public showSteps;
	public steps: IStep[] = MockSteps;
	public activeIndex: number = 1;
}

const validateWizardContent = (el: DebugElement): void => {
	const wizard = el.nativeElement;

	const header = wizard.querySelector('.l-step-wizard__header h1#test-header').textContent;
	expect(header).toContain('Test header');

	const body = wizard.querySelector('.l-step-wizard__container p#test-body').textContent;
	expect(body).toBe('Test body');

	const footer = wizard.querySelector('.l-step-wizard__footer div#test-footer').textContent;
	expect(footer).toContain('Test footer');
};

describe('WizardLayoutComponent', () => {
	let fixture: ComponentFixture<TestWrapperComponent>;
	let component: TestWrapperComponent;

	beforeEach(() => {
		TestBed.configureTestingModule({
			declarations: [
				TestWrapperComponent,
				WizardLayoutComponent,
				BreadcrumbComponent,
				BreadcrumbItemComponent,
				IconComponent,
				FilterMenuComponent,
				FilterMenuItemComponent,
				ButtonComponent,
			],
			imports: [
				StepsModule,
				ReactiveFormsModule,
				RouterTestingModule,
				TranslateModule.forRoot(),
			],
		});

		fixture = TestBed.createComponent(TestWrapperComponent);
		component = fixture.componentInstance;
	});

	describe('default state and behaviour', () => {
		it('Should render correctly', () => {
			expect(fixture).toMatchSnapshot();
		});

		it('should render the wizard without steps by default', () => {
			validateWizardContent(fixture.debugElement);

			const steps = fixture.debugElement.query(By.css('.l-step-wizard__steps'));
			expect(steps).toBeNull();
		});
	});

	describe('with steps', () => {
		it('should render the content as usual', () => {
			component.showSteps = true;
			fixture.detectChanges();

			validateWizardContent(fixture.debugElement);
		});

		it('should render the wizard with steps when showSteps is true', () => {
			component.showSteps = true;
			fixture.detectChanges();

			const steps = fixture.debugElement.query(By.css('.l-step-wizard__steps'));
			expect(steps).not.toBeNull();
		});

		it('the content of the steps should match the given steps object', () => {
			component.showSteps = true;
			fixture.detectChanges();

			const items = fixture.debugElement.queryAll(By.css('.c-step-item'));
			expect(items.length).toBe(3);

			const firstItemNumber = items[0].nativeElement.querySelector('.c-step-item__number').textContent;
			expect(parseInt(firstItemNumber, 10)).toBe(1);

			const firstItemTitle = items[0].nativeElement.querySelector('.c-step-item__title').textContent;
			expect(firstItemTitle).toContain(component.steps[0].title);

			const firstItemDescription = items[0].nativeElement.querySelector('.c-step-item__description').textContent;
			expect(firstItemDescription).toContain(component.steps[0].description);
		});
	});
});
