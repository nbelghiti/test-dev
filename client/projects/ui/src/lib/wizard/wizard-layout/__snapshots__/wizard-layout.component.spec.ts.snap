// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WizardLayoutComponent default state and behaviour Should render correctly 1`] = `
<ng-component
  activeIndex={[Function Number]}
  steps={[Function Array]}
>
  <ui-wizard-layout>
    <header
      class="l-step-wizard__header"
    >
      <div
        class="l-step-wizard__header-content l-container l-container--fluid"
      >
        
        <h1
          id="test-header"
        >
          Test header
        </h1>
      </div>
      
    </header>
    <div
      class="l-step-wizard__scroll-container"
    >
      <div
        class="l-step-wizard__container"
      >
        <div
          class="l-step-wizard__content"
        >
          
          <p
            id="test-body"
          >
            Test body
          </p>
        </div>
        
      </div>
    </div>
    <footer
      class="l-step-wizard__footer"
    >
      <div
        class="l-container l-container--fluid"
      >
        
        <div
          id="test-footer"
        >
          Test footer
        </div>
      </div>
    </footer>
  </ui-wizard-layout>
</ng-component>
`;
